subprojects/execution/src/main/java/org/gradle/internal/execution/steps/SkipEmptyWorkStep.java

```diff
@@ -86,6 +86,25 @@ public CachingResult execute(UnitOfWork work, PreviousExecutionContext context)
         }
     }
 
+    private boolean isEmptySources(ImmutableSortedMap<String, CurrentFileCollectionFingerprint> sourceFileProperties, ImmutableSet<String> propertiesRequiringIsEmptyCheck, UnitOfWork work) {
+        if (propertiesRequiringIsEmptyCheck.isEmpty()) {
+            return sourceFileProperties.values().stream()
+                .allMatch(CurrentFileCollectionFingerprint::isEmpty);
+        } else {
```

Comment:
I wonder if we want to deprecate the behavior to look inside of archive trees for doing the empty source check. It seems like a pretty niche thing and probably not worth to keep.

Reply:
That's a good question.

I can imagine one use case, where you are compiling sources downloaded from the interwebs in a ZIP format, and adding them via `zipTree()` to your `SourceTask`. In that case we'd be emitting a warning. I have absolutely no idea if this use case even exists in the wold, or how frequently it happens if it does, but it seems to me that we shouldn't break it.

IIUC, the final behavior would be to consider the presence of a ZIP backing file reason to not skip the task, even if the ZIP itself is empty, right? I think that's acceptable, though whether or not it's fully correct depends on your definition of `@SkipWhenEmpty`. Does it mean that the task should be skipped every time when its inputs are empty (so it's okay for the `@TaskAction` method not to handle an empty input?), or do you consider the feature as a best-effort performance improvement? I'd lean towards the first (let's make sure we document it), and do the change in 8.0.

As for the deprecation: I don't think we should deprecate the behavior for 7.x. How would the user fix this? I would simply leave it in the state of this PR, and then change the behavior in 8.0.

WDYT?

Reply:
In any case let's add some comments here about why we do this check now.

Reply:
Our normal suggestion for unzipping a zip is to use a `Copy` task together with a `zipTree` input. That would emit a deprecation warning if I added one here.

> I'd lean towards the first (let's make sure we document it), and do the change in 8.0.

Do you mean you lean towards the second (best-effort performance improvement)? Since for the first we don't need to make any change.

> As for the deprecation: I don't think we should deprecate the behavior for 7.x. How would the user fix this? I would simply leave it in the state of this PR, and then change the behavior in 8.0.

Sounds good to me.

Reply:
> Our normal suggestion for unzipping a zip is to use a `Copy` task together with a `zipTree` input. That would emit a deprecation warning if I added one here.
> 
> > I'd lean towards the first (let's make sure we document it), and do the change in 8.0.
> 
> Do you mean you lean towards the second (best-effort performance improvement)? Since for the first we don't need to make any change.

Yes, sorry, that's what I meant.

> > As for the deprecation: I don't think we should deprecate the behavior for 7.x. How would the user fix this? I would simply leave it in the state of this PR, and then change the behavior in 8.0.
> 
> Sounds good to me.

Let's do this then.